File: StructuralPatterns/Decorator/decorator.cpp
#include <iostream>
#include <string>

// Component Interface
class Coffee {
public:
    virtual double cost() const = 0;
    virtual ~Coffee() = default;
};

// Concrete Component
class SimpleCoffee : public Coffee {
public:
    double cost() const override {
        return 2.0; // Base cost for a simple coffee
    }
};

// Decorator
class CoffeeDecorator : public Coffee {
protected:
    Coffee* coffee;

public:
    CoffeeDecorator(Coffee* c) : coffee(c) {}

    double cost() const override {
        return coffee->cost();
    }

    virtual ~CoffeeDecorator() {
        delete coffee;
    }
};

// Concrete Decorators
class SugarDecorator : public CoffeeDecorator {
public:
    SugarDecorator(Coffee* c) : CoffeeDecorator(c) {}

    double cost() const override {
        return CoffeeDecorator::cost() + 0.5;
    }
};

class MilkDecorator : public CoffeeDecorator {
public:
    MilkDecorator(Coffee* c) : CoffeeDecorator(c) {}

    double cost() const override {
        return CoffeeDecorator::cost() + 1.0;
    }
};

class FlavorDecorator : public CoffeeDecorator {
private:
    std::string flavor;

public:
    FlavorDecorator(Coffee* c, const std::string& f) : CoffeeDecorator(c), flavor(f) {}

    double cost() const override {
        return CoffeeDecorator::cost() + 1.5;
    }

    std::string getDescription() const {
        return flavor + " " + typeid(*coffee).name();
    }

    ~FlavorDecorator() override = default;
};

// Client Code
void clientCode(const Coffee& coffee) {
    std::cout << "Cost of " << typeid(coffee).name() << ": $" << coffee.cost() << std::endl;
}

int main() {
    // Create a simple coffee
    Coffee* simpleCoffee = new SimpleCoffee();
    clientCode(*simpleCoffee);
    std::cout << "\n";

    // Decorate the simple coffee with sugar
    Coffee* sugarCoffee = new SugarDecorator(simpleCoffee);
    clientCode(*sugarCoffee);
    std::cout << "\n";

    // Decorate the sugar coffee with milk
    Coffee* sugarMilkCoffee = new MilkDecorator(sugarCoffee);
    clientCode(*sugarMilkCoffee);
    std::cout << "\n";

    // Decorate the sugar milk coffee with a flavor
    Coffee* flavoredCoffee = new FlavorDecorator(sugarMilkCoffee, "Vanilla");
    clientCode(*flavoredCoffee);

    // Cleanup
    delete simpleCoffee;
    delete sugarCoffee;
    delete sugarMilkCoffee;
    delete flavoredCoffee;

    return 0;
}

File: StructuralPatterns/Decorator/decorator.py
from abc import ABC, abstractmethod


# Component Interface
class Coffee(ABC):
    @abstractmethod
    def cost(self) -> float:
        pass


# Concrete Component
class SimpleCoffee(Coffee):
    def cost(self) -> float:
        return 2.0  # Base cost for a simple coffee


# Decorator
class CoffeeDecorator(Coffee, ABC):
    def __init__(self, coffee: Coffee):
        self._coffee = coffee

    @abstractmethod
    def cost(self) -> float:
        pass


# Concrete Decorators
class SugarDecorator(CoffeeDecorator):
    def cost(self) -> float:
        return self._coffee.cost() + 0.5


class MilkDecorator(CoffeeDecorator):
    def cost(self) -> float:
        return self._coffee.cost() + 1.0


class FlavorDecorator(CoffeeDecorator):
    def __init__(self, coffee: Coffee, flavor: str):
        super().__init__(coffee)
        self._flavor = flavor

    def cost(self) -> float:
        return self._coffee.cost() + 1.5

    def __str__(self) -> str:
        return f"{self._flavor} {self._coffee}"


def client_code(coffee: Coffee) -> None:
    print(f"Cost of {coffee}: ${coffee.cost()}")


if __name__ == "__main__":
    # Create a simple coffee
    simple_coffee = SimpleCoffee()
    client_code(simple_coffee)
    print("\n")

    # Decorate the simple coffee with sugar
    sugar_coffee = SugarDecorator(simple_coffee)
    client_code(sugar_coffee)
    print("\n")

    # Decorate the sugar coffee with milk
    sugar_milk_coffee = MilkDecorator(sugar_coffee)
    client_code(sugar_milk_coffee)
    print("\n")

    # Decorate the sugar milk coffee with a flavor
    flavored_coffee = FlavorDecorator(sugar_milk_coffee, "Vanilla")
    client_code(flavored_coffee)
